src/v/archival/archival_policy.cc

```diff
@@ -10,90 +10,118 @@
+        return {};
+    }
+    auto plog = dynamic_cast<storage::disk_log_impl*>(log->get_impl());
+    if (plog == nullptr || plog->segment_count() == 0) {
+        vlog(
+          archival_log.debug,
+          "Upload policy for {} no segments or in-memory log",
+          _ntp);
+        return {};
+    }
+    storage::segment_set& set = plog->segments();
+    const auto& ntp_conf = plog->config();
+    auto it = set.lower_bound(last_offset);
+    if (it == set.end() || (*it)->is_compacted_segment()) {
+        // Skip forward if we hit a gap or compacted segment
+        for (auto i = set.begin(); i != set.end(); i++) {
+            const auto& sg = *i;
+            if (last_offset < sg->offsets().base_offset) {
+                // Move last offset forward
+                it = i;
+                last_offset = sg->offsets().base_offset;
+                break;
             }
-            manifest::segment_meta meta{
-              .is_compacted = segment->is_compacted_segment(),
-              .size_bytes = segment->size_bytes(),
-              .base_offset = segment->offsets().base_offset,
-              .committed_offset = segment->offsets().committed_offset,
-              .is_deleted_locally = false,
-            };
-            auto path = std::filesystem::path(segment->reader().filename());
-            auto seg_name = path.filename().string();
-            // generate segment path
-            tmp.add(segment_name(seg_name), meta);
         }
-        return tmp;
     }
-
-private:
-    model::ntp _ntp;
-    model::revision_id _rev;
-};
-
-template<class PolicyImpl>
-class manifest_based_policy final
-  : public PolicyImpl
-  , public archival_policy_base {
-public:
-    manifest_based_policy(const model::ntp& ntp, const model::revision_id& rev)
-      : PolicyImpl(ntp, rev) {}
-
-    std::optional<manifest> generate_upload_set(
-      const manifest& remote, storage::log_manager& lm) final {
-        auto local = PolicyImpl::make_local_manifest(lm);
-        if (!local) {
-            return local;
-        }
-        return local->difference(remote);
+    if (it == set.end()) {
+        vlog(
+          archival_log.debug,
+          "Upload policy for {}, upload candidate is not found",
+          _ntp);
+        return {};
     }
-
-    std::optional<manifest> generate_delete_set(
-      const manifest& remote, storage::log_manager& lm) final {
-        auto local = PolicyImpl::make_local_manifest(lm);
-        if (!local) {
-            return local;
+    // Invariant: it != set.end()
+    bool closed = !(*it)->has_appender();
+    if (!closed) {
+        // Fast path, next upload candidate is not yet sealed. We may want to
+        // optimize this case because it's expected to happen pretty often. This
+        // can be done by saving weak_ptr to the segment inside the policy
+        // object. The segment must be changed to derive from
+        // ss::weakly_referencable.
+        vlog(
+          archival_log.debug,
```

Comment:
some of the debugs seem for trace. but i guess we don't have a well defined level. so is ok.